### ISSUE TITLE
Waterfox G4.0.1 build failure on Linux - undefined hidden symbol mozilla::ServoTraversalStatistics::sActive

### ISSUE THREAD

**Issue by mk-fg** [MrAlex94/Waterfox#2361]
### What happened?

Hi,

Trying to build Waterfox G4.0.1 on modern Arch Linux, had the following issue during `./mach build`:
```
36:32.21 toolkit/library/build/libxul.so
49:05.12 ld.lld: error: undefined hidden symbol: mozilla::ServoTraversalStatistics::sActive
49:05.12 >>> referenced by driver.rs:29 (servo/components/style/driver.rs:29)
49:05.12 >>>               lto.tmp:(style::driver::traverse_dom::hc3cce547b0d51e8e)
49:05.12 >>> referenced by driver.rs:29 (servo/components/style/driver.rs:29)
49:05.12 >>>               lto.tmp:(style::driver::should_report_statistics::hdab911e875bd002d)
49:05.46 clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
```

Looking at `servo/components/style/driver.rs` to work around the issue, was able to fix the build with the following trivial patch
```
--- a/servo/components/style/driver.rs
+++ b/servo/components/style/driver.rs
@@ -27,5 +27,5 @@ fn should_report_statistics() -> bool {
 #[cfg(feature = "gecko")]
 fn should_report_statistics() -> bool {
-    unsafe { crate::gecko_bindings::structs::ServoTraversalStatistics_sActive }
+    false
 }

```

There is "should_report_statistics" function for feature=servo with a similar logic shortcut just above, so maybe it's an oversight and both should be patched, or maybe my build is weird in using feature=gecko for some reason.
Might be also due to `ac_add_options --enable-linker=lld`, which I just copy-pasted from earlier PKGBUILD tbh.
Current rust version on this Arch box: `rustc 1.56.0`

Specific build script that I'm using (with mozconfig mostly copied from one in this repo):
https://github.com/mk-fg/archlinux-pkgbuilds/blob/master/waterfox/PKGBUILD

Will also paste more complete build error output in the separate box below.
Thanks!

### Reproducible?

- [ ] I have tested this on Firefox.

### Version

G4.𝑋 (Future)

### What platform are you seeing the problem on?

Linux

### Relevant log output

```shell
34:11.24 media/gmp-clearkey/0.1/libclearkey.so
34:13.04 tools/power/rapl
34:13.35 js/src/build/libjs_static.a
34:14.50 media/ffvpx/libavutil/libmozavutil.so
34:21.05 media/ffvpx/libavcodec/libmozavcodec.so
34:33.04    Compiling webrender v0.61.0 (/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.1/gfx/wr/webrender)
34:44.71    Compiling webrender_bindings v0.1.0 (/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.1/gfx/webrender_bindings)
36:30.95    Compiling gkrust v0.1.0 (/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.1/toolkit/library/rust)
36:32.03     Finished release [optimized] target(s) in 35m 35s
36:32.21 toolkit/library/build/libxul.so
49:05.12 ld.lld: error: undefined hidden symbol: mozilla::ServoTraversalStatistics::sActive
49:05.12 >>> referenced by driver.rs:29 (servo/components/style/driver.rs:29)
49:05.12 >>>               lto.tmp:(style::driver::traverse_dom::hc3cce547b0d51e8e)
49:05.12 >>> referenced by driver.rs:29 (servo/components/style/driver.rs:29)
49:05.12 >>>               lto.tmp:(style::driver::should_report_statistics::hdab911e875bd002d)
49:05.46 clang-12: error: linker command failed with exit code 1 (use -v to see invocation)
49:05.47 make[4]: *** [/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.1/config/rules.mk:545: libxul.so] Error 1
49:05.47 make[3]: *** [/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.1/config/recurse.mk:72: toolkit/library/build/target] Error 2
49:05.47 make[2]: *** [/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.1/config/recurse.mk:34: compile] Error 2
49:05.49 make[1]: *** [/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.1/config/rules.mk:355: default] Error 2
49:05.49 make: *** [client.mk:65: build] Error 2
49:05.49 0 compiler warnings present.
```

**Comment by MrAlex94** [comment]
You'll need to downgrade to rust `1.55.0` until we pull in ESR 91.3 for [this](https://hg.mozilla.org/releases/mozilla-esr91/rev/e5a5b3cbb30ea4df57567d6c1f7a25f96dce3148) patch. Or feel free to pull it in first :-)

**Comment by mk-fg** [comment]
@MrAlex94
Pretty sure you are mistaken - this patch applies perfectly in reverse, and is already included in Waterfox G4.0.1 tree as ecf25ac7 commit.
Also this build failure seem to be very different from the logs presented at https://bugzilla.mozilla.org/show_bug.cgi?id=1730397 which that commit addresses.

**Comment by MrAlex94** [comment]
Sorry, I have been looking at way too many bugs! I cannot compile with `1.56` at the moment either - but our builders are pinned to `1.55` - the same as Mozilla.

I'll report this to Mozilla as it's most likely  a bug they'll face if anyone tries to compile ESR91 with rust > `1.55`

**Comment by hawkeye116477** [comment]
> they'll face if anyone tries to compile ESR91 with rust > 1.55

Seems that's not related with rust > 1.55, cuz it works fine when building with same rust version on Ubuntu 21.10 and lld linker, but with LLVM 12.

**Comment by mk-fg** [comment]
Hm, interesting, pretty sure I've used llvm/lld 12.0.1 in my build as well.
Suspect some mozconfig option in that PKGBUILD might be triggering the issue, but just because can't think of anything more plausible, and don't know enough about rust/linking to tell for sure easily.

**Comment by mk-fg** [comment]
Error looks slightly different, but still happens for me when building G4.0.3 (without the patch above), using llvm/lld/clang 13.0.0 and rust 1.56.1 atm.

But it only happens after running `./mach python build/pgo/profileserver.py` and building with `--enable-lto=cross --enable-profile-use=cross --with-pgo-profile-path=... --with-pgo-jarlog=...` - simple build without that optimization step works without any issues.
(profiling/optimization step is inherited from earlier pkgbuild, and is also how vanilla firefox gets built on Arch)

If this is not a supported way to build waterfox (probably not the one used in official builds, at the very least), maybe issue is not relevant and can be closed.

Below is current error output when building with those pgo flags above:
```
34:57.67 media/ffvpx/libavcodec/libmozavcodec.so
47:18.89 ld.lld: error: undefined hidden symbol: mozilla::ServoTraversalStatistics::sActive
47:18.89 >>> referenced by driver.rs:29 (servo/components/style/driver.rs:29)
47:18.89 >>>               lto.tmp:(style::driver::traverse_dom::h52b64cf4c25f8f9f)
47:18.89 >>> referenced by driver.rs:29 (servo/components/style/driver.rs:29)
47:18.89 >>>               lto.tmp:(style::driver::should_report_statistics::h299d71db13c5aa70)
47:19.22 clang-13: error: linker command failed with exit code 1 (use -v to see invocation)
47:19.22 make[4]: *** [/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.3/config/rules.mk:545: libxul.so] Error 1
47:19.22 make[3]: *** [/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.3/config/recurse.mk:72: toolkit/library/build/target] Error 2
47:19.22 make[2]: *** [/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.3/config/recurse.mk:34: compile] Error 2
47:19.22 make[1]: *** [/srv/btrfs/makepkg/archlinux-pkgbuilds/waterfox/src/Waterfox-G4.0.3/config/rules.mk:355: default] Error 2
47:19.22 make: *** [client.mk:65: build] Error 2
47:19.22 0 compiler warnings present.
```

**Comment by hawkeye116477** [comment]
I tried to compile Firefox ESR 91 on OBS on Arch with same configuration as you and guess what happened. It build successfully, so apparently that means that someone from Waterfox team broke that.

**Update:** Looks like I figured out what's wrong. That code referenced by error is used by Gecko Profiler (benchmark, tool which gathers information about browser's performance) and problem is that @MrAlex94 commented out few lines in https://github.com/WaterfoxCo/Waterfox/blob/current/toolkit/moz.configure#L108, so that's why it reports errors.

[fix-profiler.patch.txt](https://github.com/WaterfoxCo/Waterfox/files/7659973/fix-profiler.patch.txt)
